paddle/fluid/distributed/fleet_executor/runtime_graph.cc

```diff
@@ -0,0 +1,190 @@
+    } else if (IsLRSched(op_role)) {
+      op_identity = static_cast<int64_t>(OpRole::kLRSched);
+    } else if (IsBackward(op_role)) {
+      op_identity = static_cast<int64_t>(OpRole::kBackward);
+    } else if (IsOptimize(op_role)) {
+      op_identity = static_cast<int64_t>(OpRole::kOptimize);
+    } else {
+      PADDLE_THROW(platform::errors::PreconditionNotMet(
+          "The op %s is None of LRSched, Forward, Backward or Optimize.",
+          op->Type()));
+    }
+    if (role_to_ops.find(op_identity) == role_to_ops.end()) {
+      role_to_ops.insert({op_identity, {}});
+    }
+    role_to_ops.at(op_identity).emplace_back(op.get());
+  }
+  int64_t cur_rank = exe_desc_.cur_rank();
+  int64_t task_id = cur_rank * num_of_functionality;
+  for (const auto& item : role_to_ops) {
+    task_nodes_.emplace_back(
+        CreateTaskNode(item.first, item.second, cur_rank, task_id));
+    ++task_id;
+  }
+}
+
+void RuntimeGraph::FakeDependence() {
+  int32_t dp_degree = exe_desc_.dp_degree();
+  int32_t pp_degree = exe_desc_.pp_degree();
+  int32_t mp_degree = exe_desc_.mp_degree();
+  int64_t cur_rank = exe_desc_.cur_rank();
+  const auto& distributed_indice =
+      DistributedIndice(dp_degree, pp_degree, mp_degree, cur_rank);
+  int64_t pp_upstream =
+      UpstreamRank(dp_degree, pp_degree, mp_degree, distributed_indice);
+  int64_t pp_downstream =
+      DownstreamRank(dp_degree, pp_degree, mp_degree, distributed_indice);
+  // forward -> lr -> backward -> optimize
+  //    |       |        |           |
+  // forward -> lr -> backward -> optimize
+  for (std::size_t i = 0; i < task_nodes_.size(); ++i) {
+    if (pp_upstream != -1) {
+      task_nodes_[i]->AddUpstreamTask(pp_upstream * num_of_functionality + i);
+    }
+    if (pp_downstream != -1) {
+      task_nodes_[i]->AddDownstreamTask(pp_downstream * num_of_functionality +
+                                        i);
+    }
+    if (i != 0) {
+      task_nodes_[i]->AddUpstreamTask(cur_rank * num_of_functionality + i - 1);
+    }
+    if (i != task_nodes_.size() - 1) {
+      task_nodes_[i]->AddDownstreamTask(cur_rank * num_of_functionality + i +
+                                        1);
+    }
+  }
+}
+
+void RuntimeGraph::AssignTaskToIntercepter() {
+  for (const auto& task : task_nodes_) {
+    int64_t intercepter_id = task->task_id();
+    if (intercepter_id_to_node_.find(intercepter_id) !=
+        intercepter_id_to_node_.end()) {
+      PADDLE_THROW(platform::errors::PreconditionNotMet(
+          "Repeated intercepter id: %d", intercepter_id));
+    }
+    intercepter_id_to_node_.insert({intercepter_id, task.get()});
+  }
+}
+
+void RuntimeGraph::FakeRuntimeInfo() {
+  int64_t nrank = exe_desc_.cluster_info().size();
+  for (int64_t i = 0; i < nrank; ++i) {
+    for (int64_t j = 0; j < num_of_functionality; ++j) {
+      int64_t intercepter_id = i * num_of_functionality + j;
+      intercepter_id_to_rank_.insert({intercepter_id, i});
```

Comment:
We have `rank` info in the task. Why not just use it when init `intercepter_id_to_node_`?